package/MDAnalysis/analysis/MSD.py

```diff
@@ -0,0 +1,398 @@
+                        temp_dict[temp_list[j]] = j
+                    self.map_atomids_to_pos_at_t[i].append(temp_dict)
+                else:
+                    pos[i].append(np.array([]))
+                    self.atomids_per_sel_at_t[i].append(set())
+                    self.map_atomids_to_pos_at_t[i].append(dict())
+
+        return
+
+    def _process_pos_data(self):
+        """Runs MSD calculation"""
+
+        n_sel = len(pos)
+        
+        calc_cutoff = self.stop_frame
+        if self.max_data:
+            calc_cutoff = len(self.universe.trajectory) - 1
+
+        # for each restart point
+        for j in range(self.start_frame, self.stop_frame+1, self.dt_restart):
+            for i in range(n_sel):  # for each selection
+
+                # storing initial set at restart
+                atoms_in_sel = self.atomids_per_sel_at_t[i][0]
+
+                # for each frame after restart
+                for ts in range(j, calc_cutoff+1):
+                    # avoid computing irrelevantly long taus
+                    if ts-j >= self.len_msd:
+                        break
+
+                    # updating restart set to exclude any atoms which have
+                    # left the selection since the restart point
+                    atoms_in_sel = atoms_in_sel.intersection(
+                        self.atomids_per_sel_at_t[i][ts-j])
+                    
+                    # find mutual atoms at times 0 and ts-j
+                    shared0 = [self.map_atomids_to_pos_at_t[i][0][k]
+                               for k in atoms_in_sel]
+                    shared = [self.map_atomids_to_pos_at_t[i][ts-j][k]
+                              for k in atoms_in_sel]
+                    
+                    # move to next restart if there's nothing to evaluate
+                    if len(shared) == 0:
+                        break  # skip to next restart
+                    
+                    self.msd[i][ts-j] = (self.msd[i][ts-j] +
+                                         squared_distance_vector(
+                                             pos[i][0][shared0],
+                                             pos[i][ts-j][shared]).sum(axis=0))
+                    self.n_samples[i][ts-j] += len(shared)
+            
+            if j % 100 == 0:
+                print("Frame: "+str(j))
+                if self.write_output:
+                    pickle.dump([self.msd/self.n_samples, self.n_samples],
+                                open(self.out_name, 'wb'))
+
+            # Update deques
+            self._update_deques(j)
+
+        return
+
+    def run(self):
+        """Calls _process_pos_data and saves final output"""
+
+        self._process_pos_data()
+
+        if self.write_output:
+            pickle.dump([self.msd/self.n_samples, self.n_samples],
+                        open(self.out_name, 'wb'))
+        # for testing purposes
+        else:
+            return [self.msd/self.n_samples, self.n_samples]
```

Comment:
To emulate the AnalysisBase API this function should always return `self` and not write any data. Writing data should be a separate function call.